File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/views/InputBarView.java
package io.euphoria.xkcd.app.impl.ui.views;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;

import io.euphoria.xkcd.app.R;
import io.euphoria.xkcd.app.impl.ui.views.MessageListAdapter.InputBarDirection;

import static io.euphoria.xkcd.app.impl.ui.UIUtils.COLOR_SENDER_LIGHTNESS;
import static io.euphoria.xkcd.app.impl.ui.UIUtils.hslToRgbInt;
import static io.euphoria.xkcd.app.impl.ui.UIUtils.nickColor;
import static io.euphoria.xkcd.app.impl.ui.UIUtils.setColoredBackground;
import static io.euphoria.xkcd.app.impl.ui.UIUtils.setEnterKeyListener;

public class InputBarView extends BaseMessageView {

    public interface NickChangeListener {

        boolean onChangeNick(InputBarView view);

    }

    public interface SubmitListener {

        boolean onSubmit(InputBarView view);

    }

    private final MarginLayoutParams defaultLayoutParams;
    private String lastNick;
    private EditText nickEntry;
    private EditText messageEntry;
    private NickChangeListener nickChangeListener;
    private SubmitListener submitListener;

    public InputBarView(Context context, AttributeSet attrs) {
        super(context, attrs);
        defaultLayoutParams = BaseMessageView.getDefaultMargins(context, attrs);
        lastNick = "";
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        nickEntry = findViewById(R.id.nick_entry);
        messageEntry = findViewById(R.id.message_entry);
        // Color nick-background with unsaturated nickname color as default
        final int defaultColor = hslToRgbInt(0, 0, COLOR_SENDER_LIGHTNESS);
        nickEntry.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // NOP
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // NOP
            }

            @Override
            public void afterTextChanged(Editable s) {
                // Recolor background
                int color = defaultColor;
                if (s.length() != 0)
                    color = nickColor(s.toString());
                setColoredBackground(nickEntry, R.drawable.bg_nick, color);
            }
        });
        setEnterKeyListener(nickEntry, EditorInfo.IME_ACTION_NEXT, new Runnable() {
            @Override
            public void run() {
                messageEntry.requestFocus();
            }
        });
        nickEntry.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (!hasFocus) doNickChange();
            }
        });
        setEnterKeyListener(messageEntry, EditorInfo.IME_ACTION_SEND, new Runnable() {
            @Override
            public void run() {
                if (submitListener == null || submitListener.onSubmit(InputBarView.this))
                    messageEntry.setText("");
            }
        });
        setColoredBackground(nickEntry, R.drawable.bg_nick, defaultColor);
    }

    @Override
    protected void updateDisplay() {
        // TODO state restoration should go here
    }

    @Override
    public void recycle() {
        super.recycle();
        // TODO state saving should go here
    }

    public String getLastNick() {
        return lastNick;
    }

    public EditText getNickEntry() {
        return nickEntry;
    }

    public EditText getMessageEntry() {
        return messageEntry;
    }

    public SubmitListener getSubmitListener() {
        return submitListener;
    }

    public void setSubmitListener(SubmitListener submitListener) {
        this.submitListener = submitListener;
    }

    public NickChangeListener getNickChangeListener() {
        return nickChangeListener;
    }

    public void setNickChangeListener(NickChangeListener nickChangeListener) {
        this.nickChangeListener = nickChangeListener;
    }

    public String getNickText() {
        return nickEntry.getText().toString();
    }

    public String getMessageText() {
        return messageEntry.getText().toString();
    }

    public void setIndent(int indent) {
        MarginLayoutParams lp = (MarginLayoutParams) getLayoutParams();
        if (lp == null) {
            lp = new MarginLayoutParams(defaultLayoutParams);
            setLayoutParams(lp);
        }
        BaseMessageView.setMarginForIndent(getContext(), lp, indent);
    }

    public boolean requestEntryFocus() {
        if (nickEntry.getText().length() == 0) {
            return nickEntry.requestFocus();
        } else {
            return messageEntry.requestFocus();
        }
    }

    public boolean mayNavigateInput(InputBarDirection dir) {
        String text = messageEntry.getText().toString();
        if (dir == InputBarDirection.ROOT || text.isEmpty()) return true;
        int index = messageEntry.getSelectionStart(), indexEnd = messageEntry.getSelectionEnd();
        return index == indexEnd && (dir == InputBarDirection.UP || dir == InputBarDirection.DOWN) &&
                !text.contains("\n");
    }

    private void doNickChange() {
        // Strip whitespace
        Editable nickEditor = nickEntry.getText();
        String newNick = nickEditor.toString();
        String trimmedNick = newNick.trim();
        if (!newNick.equals(trimmedNick)) {
            nickEditor.replace(0, newNick.length(), trimmedNick);
        }
        // Invoke listener
        if (nickChangeListener != null && !nickChangeListener.onChangeNick(this)) {
            nickEntry.setText(lastNick);
        } else {
            lastNick = nickEntry.getText().toString();
        }
    }

}

File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/views/InputBarView.kt
package io.euphoria.xkcd.app.impl.ui.views

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.View.OnFocusChangeListener
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import io.euphoria.xkcd.app.R
import io.euphoria.xkcd.app.impl.ui.UIUtils
import io.euphoria.xkcd.app.impl.ui.views.InputBarView
import io.euphoria.xkcd.app.impl.ui.views.MessageListAdapter.InputBarDirection

class InputBarView(
    context: Context,
    attrs: AttributeSet?
) : BaseMessageView(context, attrs) {
    interface NickChangeListener {
        fun onChangeNick(view: InputBarView): Boolean
    }

    interface SubmitListener {
        fun onSubmit(view: InputBarView): Boolean
    }

    private val defaultLayoutParams: MarginLayoutParams
    var lastNick: String
        private set
    var nickEntry: EditText? = null
        private set
    var messageEntry: EditText? = null
        private set
    var nickChangeListener: NickChangeListener? = null
    var submitListener: SubmitListener? = null
    override fun onFinishInflate() {
        super.onFinishInflate()
        // FIXME: work with data/view-binding to avoid so many non-null asserts
        nickEntry = findViewById(R.id.nick_entry)!!
        messageEntry = findViewById(R.id.message_entry)!!
        // Color nick-background with unsaturated nickname color as default
        val defaultColor =
            UIUtils.hslToRgbInt(0.0, 0.0, UIUtils.COLOR_SENDER_LIGHTNESS.toDouble())
        nickEntry!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(
                s: CharSequence,
                start: Int,
                count: Int,
                after: Int
            ) {
                // NOP
            }

            override fun onTextChanged(
                s: CharSequence,
                start: Int,
                before: Int,
                count: Int
            ) {
                // NOP
            }

            override fun afterTextChanged(s: Editable) {
                // Recolor background
                var color = defaultColor
                if (s.isNotEmpty()) color = UIUtils.nickColor(s.toString())
                UIUtils.setColoredBackground(nickEntry, R.drawable.bg_nick, color)
            }
        })
        UIUtils.setEnterKeyListener(
            nickEntry,
            EditorInfo.IME_ACTION_NEXT,
            Runnable { messageEntry!!.requestFocus() })
        nickEntry!!.onFocusChangeListener = OnFocusChangeListener { _, hasFocus -> if (!hasFocus) doNickChange() }
        UIUtils.setEnterKeyListener(
            messageEntry,
            EditorInfo.IME_ACTION_SEND,
            Runnable {
                if (submitListener == null || submitListener!!.onSubmit(this@InputBarView)) messageEntry!!.setText(
                    ""
                )
            })
        UIUtils.setColoredBackground(nickEntry, R.drawable.bg_nick, defaultColor)
    }

    override fun updateDisplay() {
        // TODO state restoration should go here
    }

    val nickText: String
        get() = nickEntry!!.text.toString()

    val messageText: String
        get() = messageEntry!!.text.toString()

    fun setIndent(indent: Int) {
        var lp = layoutParams as MarginLayoutParams?
        if (lp == null) {
            lp = MarginLayoutParams(defaultLayoutParams)
            layoutParams = lp
        }
        BaseMessageView.Companion.setMarginForIndent(context, lp, indent)
    }

    fun requestEntryFocus(): Boolean {
        return if (nickEntry!!.text.isEmpty()) {
            nickEntry!!.requestFocus()
        } else {
            messageEntry!!.requestFocus()
        }
    }

    fun mayNavigateInput(dir: InputBarDirection): Boolean {
        val text = messageEntry!!.text.toString()
        if (dir == InputBarDirection.ROOT || text.isEmpty()) return true
        val index = messageEntry!!.selectionStart
        val indexEnd = messageEntry!!.selectionEnd
        return index == indexEnd && (dir == InputBarDirection.UP || dir == InputBarDirection.DOWN) &&
                !text.contains("\n")
    }

    private fun doNickChange() {
        // Strip whitespace
        val nickEditor = nickEntry!!.text
        val newNick = nickEditor.toString()
        val trimmedNick = newNick.trim { it <= ' ' }
        if (newNick != trimmedNick) {
            nickEditor.replace(0, newNick.length, trimmedNick)
        }
        // Invoke listener
        if (nickChangeListener != null && !nickChangeListener!!.onChangeNick(this)) {
            nickEntry!!.setText(lastNick)
        } else {
            lastNick = nickEntry!!.text.toString()
        }
    }

    init {
        defaultLayoutParams = BaseMessageView.Companion.getDefaultMargins(context, attrs)
        lastNick = ""
    }
}
